WalletWasabi/WabiSabi/Client/AmountDecomposer.cs

```diff
@@ -211,75 +211,23 @@ public IEnumerable<Output> Decompose(IEnumerable<Money> myInputCoinEffectiveValu
 		var denoms = GetFilteredDenominations(othersInputCoinEffectiveValues.Concat(myInputCoinEffectiveValues));
 		var myInputs = myInputCoinEffectiveValues.ToArray();
 		var myInputSum = myInputs.Sum();
-		var remaining = myInputSum;
-		var remainingVsize = AvailableVsize;
 		var smallestScriptType = Math.Min(ScriptType.P2WPKH.EstimateOutputVsize(), ScriptType.Taproot.EstimateOutputVsize());
 		var maxNumberOfOutputsAllowed = Math.Min(AvailableVsize / smallestScriptType, 8); // The absolute max possible with the smallest script type.
 
 		var setCandidates = new Dictionary<int, (IEnumerable<Output> Decomposition, Money Cost)>();
 
 		// Create the most naive decomposition for starter.
-		List<Output> naiveSet = new();
-		bool end = false;
-		foreach (var denom in denoms.Where(x => x.Amount <= remaining))
-		{
-			while (denom.EffectiveCost <= remaining)
-			{
-				// We can only let this go forward if at least 2 output can be added (denom + potential change)
-				if (remaining < MinAllowedOutputAmount + ChangeFee || remainingVsize < denom.ScriptType.EstimateOutputVsize() + ChangeScriptType.EstimateOutputVsize())
-				{
-					end = true;
-					break;
-				}
-
-				naiveSet.Add(denom);
-				remaining -= denom.EffectiveCost;
-				remainingVsize -= denom.ScriptType.EstimateOutputVsize();
-
-				// Can't have more denoms than max - 1, where - 1 is to account for possible change.
-				if (naiveSet.Count >= maxNumberOfOutputsAllowed - 1)
-				{
-					end = true;
-					break;
-				}
-			}
-
-			if (end)
-			{
-				break;
-			}
-		}
-
-		var loss = Money.Zero;
-		if (remaining >= MinAllowedOutputAmount + ChangeFee)
-		{
-			naiveSet.Add(Output.FromAmount(remaining, ChangeScriptType, FeeRate));
-		}
-		else
-		{
-			// This goes to miners.
-			loss = remaining;
-		}
-
-		// This can happen when smallest denom is larger than the input sum.
-		if (naiveSet.Count == 0)
-		{
-			naiveSet.Add(Output.FromAmount(remaining, ChangeScriptType, FeeRate));
-		}
-
-		setCandidates.Add(
-			CalculateHash(naiveSet), // Create hash to ensure uniqueness.
-			(naiveSet, loss + CalculateCost(naiveSet)));
+		var naiveDecomp = CreateNaiveDecomposition(denoms, myInputSum, maxNumberOfOutputsAllowed);
+		setCandidates.Add(naiveDecomp.Key, naiveDecomp.Value);
```

Comment:
Can throw potentially.

Reply:
Good. If it creates more than one, then it better screems.

Reply:
I mean that the key is a hash set over the value, so collisions are possible while values are different. `HashSet<T>` is better for such cases because `Add` doesn't add a duplicate value, but also doesn't throw. So, you have to panic if `false` is returned.

Reply:
I mean that `setCandidates` is must be empty at this point.